### DISCUSSION TITLE
YOLO misclassifies simulated objects (Gazebo) due to COCO domain gap. Best practical fix without full retraining?

### DISCUSSION THREAD

**Discussion by fadyabirached** [ultralytics/ultralytics#25667] [Q&A] [answered]
I'm running YOLO for object-based obstacle avoidance on a TurtleBot3 in a custom Gazebo world (walls + office chairs, camera feed piped to YOLO for detection). Since the model is trained on COCO, it doesn't reliably label simulated objects correctly. A Gazebo chair often gets misclassified as whatever COCO class it's visually closest to (e.g. same silhouette/color pattern as another COCO object) rather than "chair."

For my use case I don't need class accuracy specifically. I mainly need consistent, distinguishable detections to drive navigation logic (bounding box position matters more than the exact label). But I'd like to understand the best practical way to close this gap:

Is light fine-tuning on a small set of Gazebo-rendered frames the standard approach here, or is there a lower-effort domain-adaptation trick people use for sim environments specifically?
Roughly how many labeled sim images would be reasonable for a quick fine-tune to get more reliable class assignment on 1-2 object types (chair, wall)?
Any gotchas with fine-tuning on synthetic/simulated renders vs. real photos (lighting, texture simplicity, etc.)?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fadyabirached, thank you for your interest in Ultralytics and for describing your Gazebo/TurtleBot3 use case 🚀! This is an automated response, and an Ultralytics engineer will also assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many common questions may already be answered.

For this custom training ❓ question, please provide as much information as possible, including:
- Example Gazebo-rendered images and corresponding labels
- The model, task, and exact training or inference commands
- Training logs and validation results
- Your `ultralytics` and PyTorch versions, operating system, and hardware
- Details about the simulated camera, lighting, textures, object poses, and any real-world deployment conditions
- Whether your goal is detection, segmentation, or simply stable bounding-box locations

Please also verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by SilverLucFox** [comment, marked as answer]
Good news: since you only care about consistent bounding boxes, not correct class labels, you have more low-effort options than "must fine-tune" — but fine-tuning is still the most robust fix if you want reliable class assignment specifically. Breaking down your three questions:

**1. Fine-tune vs. lower-effort tricks**

If you truly just need *consistent, distinguishable detections* for nav logic, you don't need the class label to be semantically correct at all — you need:
- Detection *presence* (something is there)
- Stable bounding box geometry (position/size correlate reliably with the physical object)

Lower-effort options before fine-tuning:
- **Class-agnostic thresholding**: ignore the label entirely, just use "any detection with confidence > X" as an obstacle. This sidesteps the whole misclassification problem if your logic doesn't need to distinguish chair vs. wall.
- **Class remapping**: keep the pretrained model as-is, but build a static lookup table mapping "whatever COCO class this Gazebo chair consistently gets called" → "chair" in your code. Since misclassification in a fixed sim world tends to be *consistent* (same lighting, same textures every episode), a wrong-but-consistent label is just as useful as a correct one for your purposes. Run a few dozen frames, tally the actual predicted classes, and hardcode the remap.
- **Classical CV fallback**: depth/proximity from the simulated camera or lidar for the "is there an obstacle" question, and just use YOLO boxes for finer localization.

Only fine-tune if you specifically need the model to say "chair" reliably (e.g., for logging, semantic decision rules, or generalizing to layouts you haven't tested).

**2. How many images for a quick fine-tune (chair + wall)**

For 1-2 classes with a fixed sim (limited texture/lighting variety), you can get usable results surprisingly cheap:
- **100-300 labeled frames per class** is a reasonable "quick and dirty" fine-tune target — enough to adapt the head/backbone to sim textures without full retraining.
- Diversity matters way more than volume here: vary camera angle, distance, occlusion, and chair pose/position in the world. 150 images from 5 different vantage points beats 500 images from one flythrough.
- Use YOLO's transfer learning workflow (load COCO-pretrained weights, freeze early backbone layers, fine-tune only last layers + head) — trains in minutes on a single GPU, sometimes even usable on CPU for that few images.
- Auto-labeling shortcut: since it's your own Gazebo world, you can get near-perfect bounding boxes for free using Gazebo's ground-truth object poses + camera projection matrix, instead of hand-labeling — turns this into a scripting problem rather than an annotation grind.

**3. Gotchas: synthetic vs. real photos**

- **Overfitting to render quirks**: Gazebo's simplified textures/lighting (flat shading, repeating textures, no real shadows/reflections) means the model can latch onto artifacts specific to your world (e.g. a specific gray gradient) rather than general "chair-ness." Fine works fine for your single sim environment, but don't expect it to transfer to a different Gazebo world or real camera feed.
- **Lighting flatness**: sim renders often lack the lighting variation that helps CNNs learn robust features — add some domain randomization (vary Gazebo world lighting/sun angle, texture color) across your training frames if you can, even cheaply, since it meaningfully improves detection stability more than volume does.
- **Aspect ratio / scale mismatch**: default TurtleBot3 camera FOV/resolution may differ meaningfully from COCO's typical image distribution — if you're seeing lots of small or distant objects, consider adjusting anchor boxes or input resolution during fine-tuning.
- **Catastrophic forgetting**: if you fine-tune too aggressively (full unfreeze, too many epochs) on a tiny sim dataset, the model can degrade on everything else — fine if you're never leaving Gazebo, but worth knowing if you later want sim-to-real transfer.
- **Sim-to-real gap remains large even after sim fine-tuning**: if there's any chance this later runs on a real TurtleBot3 with a real camera, a model fine-tuned purely on Gazebo renders can actually get *worse* on real images than the vanilla COCO model, since it's now overfit to render statistics. If sim-to-real transfer matters, mix in real COCO chair images alongside your sim frames during fine-tuning rather than training purely on synthetic data.

For your stated goal (bounding box consistency for obstacle avoidance, not class accuracy), I'd actually start with the class-agnostic/remapping approach first — it's near-zero effort and might already solve your nav problem — and only invest in fine-tuning if you find the boxes themselves are unstable (flickering, inconsistent size) rather than just mislabeled.

**Reply by fadyabirached** [reply]
Going to try the class-agnostic/remapping approach first since you're right that I don't actually need correct labels for the nav logic.  Just consistent boxes. Didn't think about using Gazebo's ground-truth poses for auto-labeling either, that's a much better idea than hand-labeling frames.

To answer your implicit question: it's staying in Gazebo for now, no real hardware in the near-term plan, so I won't worry about the sim-to-real overfitting issue, but good to know for later if that changes.

**Reply by glenn-jocher** [reply]
That sounds like the right starting point for a Gazebo-only pipeline. Ignore or remap `boxes.cls`, tune the confidence threshold for stable `boxes.xyxy`, and use Gazebo ground truth to bootstrap labels only if box stability remains insufficient. When generating a dataset, vary camera pose, distance, lighting, textures, occlusion, and object placement, and split correlated frames by scene or run rather than randomly to avoid validation leakage.

**Reply by fadyabirached** [reply]
Thanks, really appreciate the detailed pointers, will work through these.

**Reply by glenn-jocher** [reply]
You’re welcome—starting class-agnostic is sensible here. If box stability remains insufficient, fine-tune on diverse Gazebo frames using ground-truth labels and evaluate on separate scenes or runs to avoid validation leakage.

**Comment by kamalesh404** [comment]
One practical tip that pairs well with SilverLucFox's answer: if you fine-tune on Gazebo renders, Gazebo's built-in **domain randomization** plugin (`libgazebo_ros_p3d` or the `gazebo_ros_camera` material scripts) can automatically vary lighting, textures, and camera noise across training frames. This is the single most effective thing you can do with minimal effort to improve sim fine-tuning robustness -- more important than dataset size.

In practice for your TurtleBot3 setup:

1. Use `roslaunch turtlebot3_gazebo` with the Gazebo `difficulty_randomization` or world material variation plugins to generate frames under different lighting/texture conditions.
2. Export 150-200 frames across varied conditions (different sun angles, ambient light, object orientations).
3. Auto-label using Gazebo's `rosbag` + camera info + tf tree to project 3D bounding boxes into 2D, avoiding manual annotation entirely.
4. Fine-tune `yolov8n.pt` with `freeze=10` (freeze backbone, train only head) for 30-50 epochs.

The class-agnostic remapping SilverLucFox mentioned is a great zero-effort baseline though -- start there and only fine-tune if bounding box stability (not just label) is an issue.

**Reply by pderrenger** [reply]
Domain randomization is useful, but `libgazebo_ros_p3d` is not an Ultralytics domain-randomization feature, and neither 150–200 images nor `freeze=10` is universally optimal. Since class identity is secondary, start with class-agnostic navigation logic; if box stability remains poor, fine-tune pretrained weights on diverse rendered views and evaluate on a separate validation split. For a substantial COCO-to-simulation shift, also compare `freeze=10` with unfrozen training, as documented in the [Ultralytics fine-tuning guide](https://docs.ultralytics.com/guides/finetuning-guide); image diversity in lighting, textures, viewpoints, scale, and backgrounds matters more than a fixed image count.
